### PR TITLE
Instream video ad in a Protected Audience multi-seller auction demo

### PR DESCRIPTION
**Pull request by kevinkiklee**
If you have any questions or comments regarding the video demo, feel free to reply to [this discussion post](https://github.com/privacysandbox/privacy-sandbox-demos/discussions/254).

---

This PR implements instream video ad rendering in a Protected Audience multi-seller auction. 

In this demo, a pre-roll video ad has been implemented, but the same mechanism can be used to render other video ad types that use the VAST XML standard, such as mid-roll and post-roll instream video ads.

This demo shows one of the ways of handling VASTs in Protected Audience. There are other methods for Protected Audience to handle VAST XMLs, and we will publish more demos of different approaches in the future. 

This PR is dependent on the multi-seller PR: https://github.com/privacysandbox/privacy-sandbox-demos/pull/241

image

The following features have been implemented: 
* The SSP's VAST XML wraps around the DSP's VAST URI
* A unique ID is appended to the reporting URLs in the VAST XML
* The finalized VAST is messaged out of the creative iframe to the IMA SDK

To see the demo: 
* **This code has not been deployed to prod yet, but can be run locally**
* Visit https://privacy-sandbox-demos-shop.dev/items/1f45e?auctionType=multi to add the IGs
* Visit https://privacy-sandbox-demos-news.dev/?auctionType=multi to see the ads

## Description

In this demo, the DSP/SSP VASTs are transformed inside the creative, then post-messaged out of the iframe to provide the finalized VAST to the IMA SDK. 

The video ad creative that is served contains the DSP VAST URI and the SSP's endpoint URL that transforms the VAST. From inside the creative, a request to the SSP's VAST endpoint is sent and the DSP VAST URI is included in the request. On the server, SSP wraps its VAST around the DSP's VAST URI and responds back to the creative with the finalized VAST. That VAST is post-messaged from the creative iframe to the top-level frame, and passed to the IMA SDK to render the video ad. 

## Sequence

1) Buyer adds the render URLs for each SSP in the IG: 

```
const interestGroup = {
  // ...
  ads: [
    { 
      renderUrl: 'https://privacy-sandbox-demos-dsp-a.dev/html/video-ad-creative.html?sspVastUrl=https://privacy-sandbox-demos-ssp-a.dev/vast',
      metadata: {
        adType: 'video',
        seller: 'https://privacy-sandbox-demos-ssp-a.dev/'
      }
    },
    { 
      renderUrl: 'https://privacy-sandbox-demos-dsp-a.dev/html/video-ad-creative.html?sspVastUrl=https://privacy-sandbox-demos-ssp-b.dev/vast',
      metadata: {
        adType: 'video',
        seller: 'https://privacy-sandbox-demos-ssp-b.dev/'
      }
    },
  ]
}
```

DSP's code for interest group config: [Code](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/dsp-a/src/index.ts#L187)

Note that this is a temporary mechanism until `deprecatedReplaceInUrn()` for component sellers become available in M123: https://github.com/WICG/turtledove/issues/286#issuecomment-1910551260. 

After M123, the buyer can set a macro in the render URL: `https://privacy-sandbox-demos-dsp-a.dev/html/video-ad-creative.html?sspVastUrl=%%SSP_VAST_URL%%`. 

Then in the component auction config, the seller can define the following: 
```
const componentAuctionConfig = {
  // ...
  deprecatedReplaceInURN: {
    '%%SSP_VAST_URI%%': 'https://privacy-sandbox-demos-ssp-a.dev/vast',
  }
}
```

And at the render time, Protected Audience will replace the `%%SSP_VAST_URI%%` of the winning render URL with the value ('https://privacy-sandbox-demos-ssp-a.dev/vast') that the seller specified in the component auction config. 

SSP's code for component auction config: [Code](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/ssp-a/src/index.js#L144)

However, this functionality is not available for component sellers until M123, and therefore, in this demo, we are adding a render URL for each SSP as a temporary mechanism. 

2) In the bid generation function, the buyer has access to who the component auction seller is. This information is used to select the render URL for that SSP.

This step will be unnecessary once `deprecatedReplaceInURN()` is supported for component sellers.

```
function generateBid(browserSignals) {
  const {seller} = browserSignals;

  return {
    render: ads.find(ad => ad.metadata.seller.includes(seller))
  }
}
```

Bid generation code: https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/dsp-a/src/public/js/bidding-logic.js#L33

3) The winning ad is chosen and the ad is rendered. The render URL contains the location of the SSP's VAST endpoint as query params. 

The code inside the creative parses the SSP VAST URL and requests the finalized VAST from the SSP's server. The DSP's VAST URI and a unique ID is passed along in the request. 

The SSP transforms the VAST and wraps the DSP VAST URI with the SSP VAST XML, along with appending a unique ID to each reporting URL. The response back to the creative frame contains the finalized VAST XML. 

* Video creative code for calling the SSP VAST endpoint: [Code](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/dsp-a/src/public/js/video-ad-creative.js#L19)
* SSP server code for transforming the VAST: [Code](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/ssp-a/src/index.js#L184)

4) The finalized VAST XML is post-messaged out of the creative frame to the video player. 

* Video creative code for post-messaging the VAST to the ad server lib: [Code](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/dsp-a/src/public/js/video-ad-creative.js#L37)
* Ad server lib code for receiving the VAST message and passing it to IMA: [Code](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/ad-server/src/public/js/ad-server-lib.js#L105)
* IMA code for setting the VAST XML: [Code](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/news/src/public/js/video-ad-helper.js#L64)

---

### Publisher
* [Start sequential auction](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/news/src/views/index.ejs#L102)

### Ad server (top-level seller)
* [Ad server lib](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/ad-server/src/public/js/ad-server-lib.js)
* [Ad container iframe (where `runAdAuction()` is called)](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/ad-server/src/public/ad-frame.html#L35)
* [Top-level seller decision logic](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/ad-server/src/public/js/decision-logic.js#L17)

### Component SSP
* [Get contextual auction bids](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/ssp-a/src/index.js#L116)
* [Server-side VAST transform](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/ssp-a/src/index.js#L182)

### Component DSP
* [Video ad creative code](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/dsp-a/src/public/js/video-ad-creative.js)
* [Interest group generation](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/dsp-a/src/index.ts#L137)
* [Contextual auction bid generation](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/dsp-a/src/index.ts#L233)
* [Protected Audience auction bid generation](https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/dsp-a/src/public/js/bidding-logic.js#L17)

---

## Transformation

* The transformation happens on the SSP's server: https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/ssp-a/src/index.js#L184

* DSP's VAST URI: https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/dsp-a/src/public/js/video-ad-creative.js#L6

* SSP's VAST WRAPPER: https://github.com/privacysandbox/privacy-sandbox-demos/blob/dev/services/ssp-a/src/public/vast/preroll.xml

* Finalized VAST XML: 
```
<VAST version="4.0" xmlns="https://privacy-sandbox-demos-ssp-a/vast">
  <Ad id="1234" sequence="1" conditionalAd="false">
    <Wrapper followAdditionalWrappers="0" allowMultipleAds="1" fallbackOnNoAd="0">
      <AdSystem version="4.0">Privacy Sandbox</AdSystem>
      <Error>https://privacy-sandbox-demos-ssp-a.dev/report/error?auctionId=8303603</Error>
      <Impression id="Impression-ID">https://privacy-sandbox-demos-ssp-a.dev/report/impression?auctionId=8303603</Impression>
      <Creatives>
        <Creative id="1234" sequence="1" adId="1234">
          <UniversalAdId idRegistry="Ad-ID" idValue="1234">1234</UniversalAdId>
          <Linear>
            <TrackingEvents>
              <Tracking event="start">https://privacy-sandbox-demos-ssp-a.dev/report/start?auctionId=8303603</Tracking>
              <Tracking event="firstQuartile">https://privacy-sandbox-demos-ssp-a.dev/report/firstQuartile?auctionId=8303603</Tracking>
              <Tracking event="midpoint">https://privacy-sandbox-demos-ssp-a.dev/report/midpoint?auctionId=8303603</Tracking>
              <Tracking event="thirdQuartile">https://privacy-sandbox-demos-ssp-a.dev/report/thirdQuartile?auctionId=8303603</Tracking>
              <Tracking event="complete">https://privacy-sandbox-demos-ssp-a.dev/report/complete?auctionId=8303603</Tracking>
            </TrackingEvents>
            <VideoClicks>
              <ClickThrough id="shop">
                <![CDATA[https://https://privacy-sandbox-demos-shop.dev?auctionId=8303603]]>
              </ClickThrough>
            </VideoClicks>
          </Linear>
        </Creative>
      </Creatives>
      <VASTAdTagURI>
        <![CDATA[https://pubads.g.doubleclick.net/gampad/ads?iu=/21775744923/external/single_ad_samples&sz=640x480&cust_params=sample_ct%3Dlinear&ciu_szs=300x250%2C728x90&gdfp_req=1&output=vast&unviewed_position_start=1&env=vp&impl=s&correlator=]]>
      </VASTAdTagURI>
    </Wrapper>
  </Ad>
</VAST>
```

### CODE REVIEW

**Review by Seburan — COMMENTED** [review]
lgtm

File: services/ad-server/src/public/ad-frame.html [hunk 1]
```diff
@@ -46,7 +54,15 @@
           frameEl.src = auctionResult
         }
 
-        frameEl.className = 'ad'
+        if (isVideo && auctionResult) {
```
Comment (Svenmay) [inline services/ad-server/src/public/ad-frame.html:57]: Nit: Should we reverse order of elements, i.e. make it "if (auctionResult && isVideo)"? 
This would match the order as previous if-statements.

File: services/ssp-a/src/index.js [hunk 2]
```diff
@@ -45,13 +52,22 @@ app.use((req, res, next) => {
   next();
 });
 
-app.use(express.json());
```
Comment (Svenmay) [inline services/ssp-a/src/index.js:48]: Is the import "express" then still used, or could that import be removed?

File: services/ssp-b/src/index.js [hunk 3]
```diff
@@ -47,11 +54,22 @@ app.use((req, res, next) => {
 
 app.use(express.json());
```
Comment (Svenmay) [inline services/ssp-b/src/index.js:55]: This line was removed in "services/ssp-a/src/index.js", but here we keep "express", is that correct?

**Review by Svenmay — APPROVED** [review]
Added some minor comments. Nothing major, but some clarifications seem useful from my perspective.

File: services/ad-server/src/public/ad-frame.html (hunk 1 above)
Comment (kevinkiklee) [reply services/ad-server/src/public/ad-frame.html:57]: Yea, that makes it cleaner.  Switched the order.

File: services/ad-server/src/public/js/ad-server-lib.js [hunk 4]
```diff
@@ -162,20 +181,27 @@ class AdServerLib {
...
+    // Add a label
     const paragraphEl = document.createElement('p');
     paragraphEl.innerText = `${type} ad in ${
       isFencedFrame ? 'fenced frame' : 'iframe'
     }`.toUpperCase();
     paragraphEl.className = 'font-mono text-sm';
+    document.getElementById(divId).appendChild(paragraphEl);
+
+    if (type === 'image') {
+      const [width, height] = size;
```
[the remark this answers is not in the document]
Comment (kevinkiklee) [reply services/ad-server/src/public/js/ad-server-lib.js:196]: 🤯🤯🤯 Wow, I didn't know you can destructure and assign at the same time 🤯🤯🤯

Thanks a ton for this. Learned something new.

File: services/ad-server/src/public/js/ad-server-lib.js [hunk 5]
```diff
@@ -162,20 +181,27 @@ class AdServerLib {
...
+      containerFrameEl.width = width;
+      containerFrameEl.height = height;
+    } else if (type === 'video') {
+      containerFrameEl.width = 0;
```
[the remark this answers is not in the document]
Comment (kevinkiklee) [reply services/ad-server/src/public/js/ad-server-lib.js:200]: For an image ad, the image is actually rendered inside the container el iframe. 

For a video ad, what is rendered inside the container iframe is the HTML document that only contains the VAST URI.  That VAST URI is post-messaged out of the container iframe to the video player, and the ad is rendered outside of this container.  

Therefore, we hide this player since we have nothing to render in this frame, and only need to load the code that post-messages the VAST.

File: services/ssp-a/src/index.js (hunk 2 above)
Comment (kevinkiklee) [reply services/ssp-a/src/index.js:48]: Express is still used `const app = express()` in line 41.

**Comment by patmmccann** [comment]
You should add to this demo or make an independent demo clarifying if publisher video content runs after the video ad. Video ad is quite vague; most people refer to ads with content, but this appears to be in banner video

**Comment by kevinkiklee** [comment]
> You should add to this demo or make an independent demo showing publisher video content running after the video ad. Video ad is quite vague; most people refer to ads with content, but this appears to be in banner video. 

Hi there!  Yes, this is a preroll ad and the publisher's video will play after the preroll ad finishes. This is where the publisher's video is: https://github.com/privacysandbox/privacy-sandbox-demos/blob/video-ad/services/news/src/views/index.ejs#L42

> Am I reading this accurately? Can the postmessage out of the frame support pre-roll? If so lets make that crystal.

Yes, the finalized VAST XML (DSP+SSP VASTs) is post-messaged out of the iframe to the top-level frame. 

* This is where the VAST XML is post-messaged out of a Protected Audience ad iframe: https://github.com/privacysandbox/privacy-sandbox-demos/blob/video-ad/services/dsp-a/src/public/js/video-ad-creative.js#L37
* And this is where that VAST is received : https://github.com/privacysandbox/privacy-sandbox-demos/blob/video-ad/services/ad-server/src/public/js/ad-server-lib.js#L105
* And this is where the VAST is set for IMA: https://github.com/privacysandbox/privacy-sandbox-demos/blob/video-ad/services/news/src/public/js/video-ad-helper.js#L62

> The piece we're curious to see is analogous to this https://googleads.github.io/googleads-ima-html5/vsi/ . 

I've posted the finalized VAST XML that is post-messaged out of the iframe in the descriptions. You can copy-paste that XML into that VAST XML text box on the IMA inspector page to see the preroll ad. 

---

Also, we are in the process of trying to deploy this ASAP, and we are targeting sometime next week.  However, you can pull down the branch and play around with it yourself locally.

**Comment by michaelkleber** [comment]
Hey @kevinkiklee This is a great demo!

I think we should also put together an example of this same idea but where the SSP (component seller) does the work, and the DSP buyer's ad server just serves the same XML bytes that they have always served.

Seems like the easiest way to support that would be for the renderURL itself to be the thing that changes.  Instead of the Interest Group carrying around a renderURL like `buyer.com/my-vast-here.xml`, they could change to a renderURL like `component-seller.com/vast-wrapper.html?creative=buyer.com%2Dmy-vast-here.xml`.

**Comment by patmmccann** [comment]
@kevinkiklee thanks for clarifying! I recommend you re-title the PR 'Video Pre-roll ad...'

Nice work!

**Comment by patmmccann** [comment]
cc @karimmouraa

**Review by siddharth-sahoo — APPROVED** [review]

**Comment by kevinkiklee** [comment]
> I think we should also put together an example of this same idea but where the SSP (component seller) does the work, and the DSP buyer's ad server just serves the same XML bytes that they have always served.

@michaelkleber Sounds good. I'll get that demo started once I wrap up the open PRs. 

### SSP-supplied Render URL approach

1. The DSP sets the following render URL in the IG: 

```
const interestGroup = {
  // ...
  ads: [
    { 
      renderUrl: 'https://privacy-sandbox-demos-ssp-a.dev/video-ad.html?dspVastUri=https://privacy-sandbox-demos-dsp-a.dev/preroll.xml',
      // ...
    }
  ]
}

```
The render URL points to the SSP’s video ad serving endpoint, and the DSP’s VAST URI is added as query params. 

The SSP is now responsible for serving the actual ad that is rendered inside the iframe, and it also contains the finalized VAST XMLs (SSP+DSP VAST XMLs). 

2. When that ad wins the auction, the browser makes a request to the render URL which is the SSP's ad serving endpoint `/video-ad.html` with the DSP's VAST URI set in the query params
3. The SSP’s `/video-ad.html` endpoint responds with the HTML document that contains the finalized VAST XML
4. SSP’s HTML document is rendered in the ad iframe
5. The finalized VAST XML is post-messaged out of the creative frame to the video player

**Comment by michaelkleber** [comment]
> 3. The SSP’s `/video-ad.html` endpoint responds with the HTML document that contains the finalized VAST XML

That seems like a very reasonable implementation, though another choice would be for the SSP's video-ad.html code to fetch the VAST XML directly from the browser and do the wrapping browser-side.  I don't know which approach SSPs would like better; it seems to me like they would both work.

**Comment by kevinkiklee** [comment]
> That seems like a very reasonable implementation, though another choice would be for the SSP's video-ad.html code to fetch the VAST XML directly from the browser and do the wrapping browser-side. 

Yea, client-side transformation is definitely doable. For the demo, wrapping on the client-side would be easier to implement, so I'll just go with that now, but I can easily switch the implementation if I receive feedback that SSPs prefer server-side processing.

**Comment by adamleslie** [comment]
Hi, how does this support video fallback? https://support.google.com/admanager/answer/3007370?hl=en

This is a highly used feature in video advertising for publishers

**Comment by kevinkiklee** [comment]
> Hi, how does this support video fallback? https://support.google.com/admanager/answer/3007370?hl=en
> 
> This is a highly used feature in video advertising for publishers

@adamleslie Hi there! The goal of this demo is to show one way to get VAST output from a Protected Audience auction, and the VAST XML can include the fallback video ad information: https://support.google.com/admanager/answer/4517679#:~:text=Fallback%2C%20also%20known%20as%20%22waterfall%22

Then, the SDK implements the fallback logic based on this VAST, like the GAM video fallback flow you linked to.

**Comment by adamleslie** [comment]
Hey, thank you. Just to confirm; the VAST output from a Protected Audience top-level auction would come back as essentially a stitched playlist of multiple VAST XML documents from the various Protected Audience component auctions.

The playlist of multiple VAST XML documents would be handled by the renderer, e.g. the IMA SDK - as is done today.

Is that correct?

**Comment by michaelkleber** [comment]
Hi @adamleslie: No, the Protected Audience auction only allows a winner to come from a single Interest Group buying through a single component auction.  Since this demo is based on using the Protected Audience API in the form that is already available in Chrome today, it cannot combine things across multiple IGs.

**Comment by rdgordon-index** [comment]
Curious about the transition from an opaque URN to VAST XML that can be loaded by a player -- neither `XMLHTTPRequest` nor `fetch()` seem to be able to work with URNs.

**Comment by adamleslie** [comment]
Ok, so lots of demand goes into the top-level auction; only one demand source comes out of the top-level auction - this would be an issue if that understanding is correct.

Summarising, moving to this model may have notable impacts to video advertising fill for Publishers - up to say 50% less fill and thus revenues.

**Detail:**
Current video advertising operations behave differently from display due to the expected functionality of VAST. This is to say that even in a scenario where you run auctions before trying to render (say through Prebid), buyers commonly run additional logic (say brand safety checks) at the point of time when the player tries to render the underlying mediafile from the VAST XML document. Today, this is not so big of an issue as most Publishers tend to populate several VAST XMLs into a playlist (say 5 per video opportunity) for the player to mediate through using VAST errors; however in a scenario where this is reduced to one there is a greater chance that it errors and causes a missed video ad opportunity.

_Furthermore,_
Considering the above point and integration with wider demand (say direct advertising campaigns which are VAST redirects or holding group multi deal IDs), it's not obvious how or when these PAAPI components from the top level auction are included into the render process (read: VAST playlist which typically is the response object to the player from the ad server) for the IMA SDK.

_In short,_
- It would be good to explore including more components from the top-level auction understanding the current behaviours of advertising in practice
- It would be good to understand in greater detail how this integrates into the wider advertising operations, specifically technologies such as the Google IMA SDK

If you have any further questions on the points I've made - let me know!

**Comment by kevinkiklee** [comment]
Hi @adamleslie. We have started a discussion post for the video ads here: https://github.com/privacysandbox/privacy-sandbox-demos/discussions/254. 

I've migrated your comment and will respond there.

**Comment by kevinkiklee** [comment]
If anyone has any comments or questions regarding the video ad demo, please use the discussion post: [Demo: Instream video ad in a Protected Audience sequential auction setup](https://github.com/privacysandbox/privacy-sandbox-demos/discussions/254)

**Review by mehdiamiri11 — COMMENTED** [review]
AGPL3.O+``
